driver-core/src/main/java/com/datastax/driver/core/CodecRegistry.java

```diff
@@ -0,0 +1,638 @@
+     * given Java type, that codec may be returned if it is found first
+     * in the registry, <em>even if another codec is a better match</em>.
+     * <p>
+     * Codecs returned by this method are cached.
+     *
+     * @param cqlType The {@link DataType CQL type} the codec should accept; must not be {@code null}.
+     * @param javaType The {@link TypeToken Java type} the codec should accept; must not be {@code null}.
+     * @return A suitable codec.
+     * @throws CodecNotFoundException if a suitable codec cannot be found.
+     */
+    public <T> TypeCodec<T> codecFor(DataType cqlType, TypeToken<T> javaType) throws CodecNotFoundException {
+        checkNotNull(cqlType, "Parameter cqlType cannot be null");
+        checkNotNull(javaType, "Parameter javaType cannot be null");
+        return lookupCache(cqlType, javaType);
+    }
+
+    /**
+     * Returns a {@link TypeCodec codec} that accepts the given {@link DataType CQL type}
+     * and the given value.
+     * <p>
+     * This method takes an actual Java object and tries to locate a suitable codec for it.
+     * For this reason, codecs must perform a {@link TypeCodec#accepts(Object) "manual" inspection}
+     * of the object to determine if they can accept it or not, which, depending on the implementations,
+     * can be an expensive operation; besides, the resulting codec cannot be cached.
+     * Therefore there might be a performance penalty when using this method.
+     * <p>
+     * Codecs returned by this method <em>may</em> be cached.
+     *
+     * @param cqlType The {@link DataType CQL type} the codec should accept; must not be {@code null}.
+     * @param value The value the codec should accept; must not be {@code null}.
+     * @return A suitable codec.
+     * @throws CodecNotFoundException if a suitable codec cannot be found.
+     */
+    public <T> TypeCodec<T> codecFor(DataType cqlType, T value) {
+        checkNotNull(cqlType, "Parameter cqlType cannot be null");
+        checkNotNull(value, "Parameter value cannot be null");
+        TypeToken<T> javaType = guessJavaType(value);
+        if(javaType != null)
+            // if we are sure we have the right Java type, hit the cache
+            return lookupCache(cqlType, javaType);
+        else
+            // otherwise, iterate over the codecs and rely on TypeCodec.accepts()
+            return findCodec(cqlType, value);
+    }
+
+    private void doRegisterCodec(TypeCodec<?> codec){
+        for (TypeCodec<?> existing : codecs) {
+            if(existing.getJavaType().equals(codec.getJavaType())) {
+                logger.warn("Attempting to register a codec {} whose Java type is already handled by another codec: {}; "
+                        + "the former will be chosen over the latter if the registry cannot determine which one fits the best. "
+                        + "This usually denotes a misconfiguration.",
+                    codec, existing);
```

Comment:
I'm not sure I agree: this will only be a problem in situations where the CQL type is unknown (SimpleStatement and BuiltStatement if I'm correct). I can the case for having two CQL types mapping to the same Java type, provided that you explicitly avoid these situations.
In a way this is already the case with some of the primitive types: bigint and counter (Long), uuid and timeuuid (UUID).
I think a warning log is excessive.